### ISSUE TITLE
Pull request branch not updated when the diff is empty, even though there are changes in the tree

### ISSUE THREAD

**Issue by oakkitten** [peter-evans/create-pull-request#1191]
I suppose that this is working as intended, as the readme says,

> If a pull request already exists and there are no further changes (i.e. no diff with the current pull request branch) then the action exits silently.

But this can lean to a situation where the PR branch (and hence the PR) is not updated even though there are changes in the tree, such as when the main branch was force pushed after rebase. I even ran into a situation with an empty diff and GitHub saying “This branch has conflicts that must be resolved”. While this problem is rare, probably extremely rare, perhaps there's a way to to update PR branch even when the diff is empty?

If you want to reproduce this behavior, force-pushing main branch without changes while PR is pending should trigger the issue if the workflow is set up in a more or less ordinary way. Here's [one workflow run](https://github.com/oakkitten/anki-wallpaper/runs/6676894322?check_suite_focus=true#step:8:488) with this problem.

**Comment by peter-evans** [comment]
Hi @oakkitten 

What state is the PR in after you force push the main branch? Can it still be merged? Or is it complaining that there is a conflict?

If this is a rare problem, my inclination would be to suggest dealing with it by closing the PR and letting your workflow reopen it when it next runs. Note that if you don't want to wait for it to rerun you could add the `workflow_dispatch` event so you can trigger it manually from the UI.

**Comment by peter-evans** [comment]
> If you want to reproduce this behavior, force-pushing main branch without changes while PR is pending should trigger the issue if the workflow is set up in a more or less ordinary way.

When I have some time I'll test this scenario. The action only checks if there is a diff, so changes to commits after a force push on the target branch are probably being ignored. If such changes are leaving the PR in a state where it cannot be merged then I think that is not good behaviour, and perhaps there might be a way to detect that and update the PR by rebasing on to its target.

**Comment by highb** [comment]
Chiming in here to say that I've been experiencing a similar issue on my [kustomize-everything/action-env-build-and-deploy Github action](https://github.com/kustomize-everything/action-env-build-and-deploy) that uses `peter-evans/create-pull-request@v4`. I found that I had to do a couple of things to get the PR creation action to work the way I was expecting every time:
1. Ensure that the branch that I am passing as `branch` to the action exists and has been pushed to the remote.
2. DO NOT commit my changes before calling the action. I found that any change that I attempted to make and commit to a branch and then open a PR with using the action would result in the action not detecting any change, which would then result in auto-closing the PR.

[Here's the diff between a usage of the action that wasn't working for me and one that is](https://github.com/kustomize-everything/action-env-build-and-deploy/compare/v1.1.0...v1.1.2), if that is helpful to anyone.

**Comment by peter-evans** [comment]
@oakkitten I tried to reproduce the issue.

1. Triggered a workflow that opens a PR with a change to merge to `main`
2. Modified the commit message of the HEAD commit of `main` and force pushed. ("force-pushing main branch without changes")
3. Trigger the workflow again to update the PR

After the force push on step 2, the diff of the PR is wrong. I guess this expected and just how git calculates the diff in this situation. However, after the workflow runs in step 3, the PR updates (rebases itself) and corrects the diff.

So I'm not seeing any issue here. Am I trying to reproduce this incorrectly?

**Comment by bgilbert** [comment]
[This demo repo](https://github.com/bgilbert/1191-demo) has some workflows that should reproduce the problem:

1. Fork the repo.
2. Run the "PR" workflow in the fork to create a PR.
3. Run the "Rewrite last commit" workflow to force-push the `main` branch.
4. Run the "PR" workflow again.  It won't rebase the `update` branch onto the new tip of `main`.

**Comment by peter-evans** [comment]
@bgilbert Thank you for creating that demo repo. It was very useful!

I've created a fix for this issue [here](https://github.com/peter-evans/create-pull-request/pull/1307). I also tested it using the demo repo @bgilbert created. You can see that [this PR](https://github.com/peter-evans/1191-demo/pull/1) got updated after the force push of `main` to a new commit.

I'm about to go away for ~10 days so I don't want to merge this now, just in case it causes some unexpected issue and I need to roll it back. So you can either wait, or you can use the feature branch temporarily and update to `v4` again after I release it.

If you want to use the feature branch:
```yml
      - uses: peter-evans/create-pull-request@base-force-push
```

**Comment by peter-evans** [comment]
Released as `v4.2.1` / `v4`

Please let me know if/when it is safe to delete the `base-force-push` feature branch.

**Comment by bgilbert** [comment]
Good to delete from my side.  Thanks for the fix!
